src/main/java/com/faforever/client/teammatchmaking/TeamMatchmakingController.java

```diff
@@ -0,0 +1,325 @@
+    for(int i = 0; i < members.size(); i++) {
+      PartyMemberItemController controller = uiService.loadFxml("theme/play/teammatchmaking/matchmaking_member_card.fxml");
+      controller.setMember(members.get(i));
+      if (members.size() == 1) {
+        partyMemberPane.add(controller.getRoot(), 0, 0, 2, 1);
+      } else {
+        partyMemberPane.add(controller.getRoot(), i % 2, i / 2);
+      }
+    }
+  }
+
+  @Override
+  public Node getRoot() {
+    return teamMatchmakingRoot;
+  }
+
+  public void onInvitePlayerButtonClicked(ActionEvent actionEvent) {
+    InvitePlayerController invitePlayerController = uiService.loadFxml("theme/play/teammatchmaking/matchmaking_invite_player.fxml");
+    Pane root = invitePlayerController.getRoot();
+    uiService.showInDialog(teamMatchmakingRoot, root, i18n.get("teammatchmaking.invitePlayer"));
+  }
+
+  public void onLeavePartyButtonClicked(ActionEvent actionEvent) {
+    teamMatchmakingService.leaveParty();
+  }
+
+  public void onFactionButtonClicked(ActionEvent actionEvent) {
+
+    if (!uefButton.isSelected() && !aeonButton.isSelected() && !cybranButton.isSelected() && !seraphimButton.isSelected()) {
+      selectFactionsBasedOnParty();
+      return;
+    }
+
+    List<Faction> factions = new ArrayList<>();
+    if (uefButton.isSelected()) {
+      factions.add(Faction.UEF);
+    }
+    if (aeonButton.isSelected()) {
+      factions.add(Faction.AEON);
+    }
+    if (cybranButton.isSelected()) {
+      factions.add(Faction.CYBRAN);
+    }
+    if (seraphimButton.isSelected()) {
+      factions.add(Faction.SERAPHIM);
+    }
+
+    teamMatchmakingService.sendFactionSelection(factions);
+
+    refreshingLabel.setVisible(true);
+  }
+
+  private void selectFactionsBasedOnParty() {
+    uefButton.setSelected(isFactionSelectedInParty(Faction.UEF));
+    aeonButton.setSelected(isFactionSelectedInParty(Faction.AEON));
+    cybranButton.setSelected(isFactionSelectedInParty(Faction.CYBRAN));
+    seraphimButton.setSelected(isFactionSelectedInParty(Faction.SERAPHIM));
+  }
+
+  private boolean isFactionSelectedInParty(Faction faction) {
+    return teamMatchmakingService.getParty().getMembers().stream()
+        .anyMatch(m -> m.getPlayer().getId() == player.getId() && m.getFactions().contains(faction));
+  }
+
+  private void createChannelTab(String channelName) {
+    matchmakingChatController = uiService.loadFxml("theme/play/teammatchmaking/matchmaking_chat.fxml");
+    matchmakingChatController.setChannel(channelName);
+    Platform.runLater(() -> {
+      chatTabPane.getTabs().clear();
+      chatTabPane.getTabs().add(matchmakingChatController.getRoot());
+    });
+  }
+
+  @Subscribe
+  public void onChatMessage(ChatMessageEvent event) {
+    Platform.runLater(() -> {
+      ChatMessage message = event.getMessage();
+      if (message.getSource().equals(matchmakingChatController.getReceiver())) {
+        matchmakingChatController.onChatMessage(message);
```

Comment:
Since there might be a lot of messages received here. Only execute this very line on the UI  Thread no need to switch to the UI thread to check if the message is releavnt => 99% less execution on the UI thread for this action